backend_sndio/backend_sndio.c

```diff
@@ -0,0 +1,220 @@
+
+	if (pbe->arg)
+		free(pbe->arg);
+}
+
+static void
+sndio_onmove_cb(void* addr, int delta)
+{
+	struct voss_backend *pbe = addr;
+	/* Convert delta from frames to bytes using params assumed below */
+	get_sio_ctx(pbe)->realpos += (delta * /*bps=*/2 * /*pchan=*/2);
+}
+
+static int sndio_get_signedness(int *fmt) {
+	int s_fmt = *fmt & (VPREFERRED_SLE_AFMT | VPREFERRED_SBE_AFMT);
+	if (s_fmt) {
+		*fmt = s_fmt;
+		return 1;
+	}
+	*fmt = *fmt & (VPREFERRED_ULE_AFMT | VPREFERRED_UBE_AFMT);
+	return 0;
+}
+
+static int sndio_get_endianness_is_le(int *fmt) {
+	int le_fmt = *fmt & (VPREFERRED_SLE_AFMT | VPREFERRED_ULE_AFMT);
+	if (le_fmt) {
+		*fmt = le_fmt;
+		return 1;
+	}
+	*fmt = *fmt & (VPREFERRED_SBE_AFMT | VPREFERRED_UBE_AFMT);
+	return 0;
+}
+
+static int sndio_get_bits(int *fmt) {
+	if (*fmt & (AFMT_S16_LE | AFMT_U16_LE | AFMT_S16_BE | AFMT_U16_BE)) {
+		return 16;
+	}
+	if (*fmt & (AFMT_S24_LE | AFMT_U24_LE | AFMT_S24_BE | AFMT_U24_BE)) {
+		return 24;
+	}
+	if (*fmt & (AFMT_S32_LE | AFMT_U32_LE | AFMT_S32_BE | AFMT_U32_BE)) {
+		return 32;
+	}
+	if (*fmt & (AFMT_S8 | AFMT_U8)) {
+		return 8;
+	}
+	return -1;
+}
+
+static int
+sndio_open(struct voss_backend *pbe, const char *devname,
+    int samplerate, int bufsize, int *pchannels, int *pformat)
+{
+    const char *sndio_name = devname + 6;
+
+	int sig = sndio_get_signedness(pformat);
+	int le  = sndio_get_endianness_is_le(pformat);
+	int bits = sndio_get_bits(pformat);
+
+	if (bits == -1) {
+		warn("unsupported format precision");
+		return (-1);
+	}
+
+    struct sio_hdl* hdl =
+        sio_open(sndio_name, SIO_PLAY, 0);
+    if (hdl == 0) {
+        warn("sndio: failed to open device");
+        return (-1);
+    }
+
+    struct sio_par par;
+    sio_initpar(&par);
+	par.pchan = *pchannels;
+	par.sig = sig;
+	par.bits = bits;
+	par.bps = SIO_BPS(bits);
+	par.le = le;
+	par.rate = samplerate;
+	par.appbufsz = bufsize;
+	par.xrun = SIO_SYNC;
+	if (!sio_setpar(hdl, &par))
+		errx(1, "internal error, sio_setpar() failed");
+	if (!sio_getpar(hdl, &par))
+		errx(1, "internal error, sio_getpar() failed");
+	if (par.pchan != *pchannels)
+		errx(1, "couldn't set number of channels");
+	if (par.sig != sig || par.bits != bits || par.le != le)
+		errx(1, "couldn't set format");
+	if (par.bits != bits)
+		errx(1, "couldn't set precision");
+	if (par.rate < samplerate * 995 / 1000 ||
+			par.rate > samplerate * 1005 / 1000)
+		errx(1, "couldn't set rate");
+	if (par.xrun != SIO_SYNC)
+		errx(1, "couldn't set xun policy");
+
+	/* Save the device handle in a context struct */
+	struct sio_ctx* ctx = malloc(sizeof(struct sio_ctx));
+	memset(ctx, 0, sizeof(struct sio_ctx));
+	ctx->hdl= hdl;
+	pbe->arg = ctx;
+
+	/* Set up a callback to help compute delay */
+	sio_onmove(hdl, sndio_onmove_cb, pbe);
+
+	/* Start the device.*/
+	if (!sio_start(hdl))
+		errx(1, "couldn't start device");
+
+	return (0);
+}
+
+static int
+sndio_play_transfer(struct voss_backend *pbe, void *ptr, int len)
```

Comment:
Would it be difficult to add support for recording aswell?

Reply:
My hope is that it would be easy for one to extend this to support recording, but I'm not set up to test it. My hardware and use cases are currently playback-only.

Reply:
I can fix the recording part. Looks easy enough.